File: test/emu/cdchost.mjs
// The PC's side of a composite USB device with CDC serial ports, on the
// emulated RP2040's device controller (rp2040js's RPUSBController): rp2040js's
// USBCDC (src/usb/cdc.ts), which knows one port, grown to every CDC function
// the configuration descriptor has. The system card has two (the sysctl
// protocol and the console, doc/proposals/usb-console.md).
//
// Enumeration is USBCDC's, step for step: address 1, the configuration
// descriptor (9 bytes, then all of it), configuration 1, then DTR|RTS on
// port 0's communication interface and onDeviceConnected. After that, a
// control request (setLines) waits for the previous one's status stage.
// Port p: its data comes to ports[p].onSerialData; sendSerialByte(b, p)
// queues bytes for it (512-byte FIFO each, as USBCDC's).
//
// The C++ port (emu/rp2040/src/usb/cdchost.cpp) behaves identically: the
// system card tests (SYS-006) compare the two (EMU-006).

import path from 'node:path';
import { SDK } from './rp2040emu.mjs';

const rp = await import(path.join(SDK, 'rp2040js/dist/esm/index.js'));
const { FIFO } = await import(path.join(SDK, 'rp2040js/dist/esm/utils/fifo.js'));

const CDC_REQUEST_SET_CONTROL_LINE_STATE = 0x22;
const CDC_DTR = 1, CDC_RTS = 2;
const CDC_COMM_CLASS = 2, CDC_DATA_CLASS = 10;
const ENDPOINT_BULK = 2;
const CONFIGURATION_DESCRIPTOR_SIZE = 9;

// [{ comm, in, out }] per CDC function, in descriptor order: a data interface
// with two bulk endpoints after its communication interface
export function extractPorts(d) {
  const ports = [];
  let comm = -1, data = null;
  for (let i = 0; i < d.length;) {
    const len = d[i];
    if (len < 2 || d.length < i + len) break;
    const type = d[i + 1];
    if (type === rp.DescriptorType.Interface && len === 9) {
      data = null;
      if (d[i + 5] === CDC_COMM_CLASS) comm = d[i + 2];
      else if (d[i + 5] === CDC_DATA_CLASS && d[i + 4] === 2) {
        data = { comm, in: -1, out: -1 };
        ports.push(data);
      }
    }
    if (data && type === rp.DescriptorType.Endpoint && len === 7 && (d[i + 3] & 3) === ENDPOINT_BULK) {
      if (d[i + 2] & 0x80) data.in = d[i + 2] & 0xf;
      else data.out = d[i + 2] & 0xf;
    }
    i += len;
  }
  return ports;
}

export class CdcHost {
  constructor(usb, nports = 2) {
    this.usb = usb;
    this.ports = Array.from({ length: nports }, () => ({ txFIFO: new FIFO(512), onSerialData: null, comm: -1, in: -1, out: -1 }));
    this.initialized = false;
    this.descriptorsSize = null;
    this.descriptors = [];
    this.pending = [];          // control requests waiting for the one in progress
    this.busy = false;
    this.onDeviceConnected = null;
    usb.onUSBEnabled = () => usb.resetDevice();
    usb.onResetReceived = () => usb.sendSetupPacket(rp.setDeviceAddressPacket(1));
    usb.onEndpointWrite = (ep, buf) => {
      if (ep === 0 && buf.length === 0) {
        if (this.descriptorsSize == null) {
          usb.sendSetupPacket(rp.getDescriptorPacket(rp.DescriptorType.Configration, CONFIGURATION_DESCRIPTOR_SIZE));
        } else if (!this.initialized) {
          // configured: port 0 opens, as USBCDC's one port does
          this.initialized = true;
          this.busy = false;
          this.setLines(0, CDC_DTR | CDC_RTS);
          this.onDeviceConnected?.();
        } else {
          // a control request's status stage: the next may go
          this.busy = false;
          this.next();
        }
      }
      if (ep === 0 && buf.length > 1) {
        if (buf.length === CONFIGURATION_DESCRIPTOR_SIZE && buf[1] === rp.DescriptorType.Configration &&
            this.descriptorsSize == null) {
          this.descriptorsSize = (buf[3] << 8) | buf[2];
          usb.sendSetupPacket(rp.getDescriptorPacket(rp.DescriptorType.Configration, this.descriptorsSize));
        } else if (this.descriptorsSize != null && this.descriptors.length < this.descriptorsSize) {
          this.descriptors.push(...buf);
        }
        if (this.descriptorsSize === this.descriptors.length) {
          extractPorts(this.descriptors).slice(0, this.ports.length).forEach((f, i) => Object.assign(this.ports[i], f));
          usb.sendSetupPacket(rp.setDeviceConfigurationPacket(1));
        }
      }
      if (ep !== 0) {
        const p = this.ports.find((q) => q.in === ep);
        p?.onSerialData?.(buf);
      }
    };
    usb.onEndpointRead = (ep, size) => {
      const p = this.ports.find((q) => q.out === ep);
      if (!p) return;
      const b = new Uint8Array(Math.min(size, p.txFIFO.itemCount));
      for (let i = 0; i < b.length; i++) b[i] = p.txFIFO.pull();
      usb.endpointReadDone(ep, b);
    };
  }

  // the number of CDC functions the device has (known once configured)
  get portCount() {
    return this.ports.filter((p) => p.in >= 0).length;
  }

  // SET_CONTROL_LINE_STATE on port p's communication interface: DTR is
  // "a terminal has the port open"
  setLines(p, value) {
    this.pending.push([p, value]);
    this.next();
  }

  open(p, on) {
    this.setLines(p, on ? CDC_DTR | CDC_RTS : 0);
  }

  next() {
    if (this.busy || !this.initialized || !this.pending.length) return;
    this.busy = true;
    const [p, value] = this.pending.shift();
    this.usb.sendSetupPacket(rp.createSetupPacket({
      dataDirection: rp.DataDirection.HostToDevice, type: rp.SetupType.Class, recipient: rp.SetupRecipient.Interface,
      bRequest: CDC_REQUEST_SET_CONTROL_LINE_STATE, wValue: value, wIndex: this.ports[p].comm, wLength: 0,
    }));
  }

  sendSerialByte(b, p = 0) {
    this.ports[p].txFIFO.push(b);
  }
}

File: emu/rp2040/src/usb/cdchost.cpp
// Port of test/emu/cdchost.mjs
#include "cdchost.h"

#include <algorithm>

#include "../peripherals/usb.h"
#include "setup.h"

namespace rp2040js {

static constexpr uint32_t CDC_REQUEST_SET_CONTROL_LINE_STATE = 0x22;
static constexpr uint32_t CDC_DTR = 1, CDC_RTS = 2;
static constexpr uint32_t CDC_COMM_CLASS = 2, CDC_DATA_CLASS = 10;
static constexpr uint32_t ENDPOINT_BULK = 2;
static constexpr uint32_t CONFIGURATION_DESCRIPTOR_SIZE = 9;

std::vector<CdcHost::Function> CdcHost::extractPorts(const std::vector<uint8_t> &d) {
  std::vector<Function> ports;
  int32_t comm = -1;
  bool data = false;
  for (size_t i = 0; i < d.size();) {
    const uint32_t len = d[i];
    if (len < 2 || d.size() < i + len) break;
    const uint32_t type = d[i + 1];
    if (type == static_cast<uint32_t>(DescriptorType::Interface) && len == 9) {
      data = false;
      if (d[i + 5] == CDC_COMM_CLASS) {
        comm = d[i + 2];
      } else if (d[i + 5] == CDC_DATA_CLASS && d[i + 4] == 2) {
        data = true;
        ports.push_back({comm, -1, -1});
      }
    }
    if (data && type == static_cast<uint32_t>(DescriptorType::Endpoint) && len == 7 &&
        (d[i + 3] & 3) == ENDPOINT_BULK) {
      if (d[i + 2] & 0x80)
        ports.back().in = d[i + 2] & 0xf;
      else
        ports.back().out = d[i + 2] & 0xf;
    }
    i += len;
  }
  return ports;
}

CdcHost::CdcHost(RPUSBController &usb, size_t nports) : ports(nports), usb(usb) {
  usb.onUSBEnabled = [this] { this->usb.resetDevice(); };
  usb.onResetReceived = [this] { this->usb.sendSetupPacket(setDeviceAddressPacket(1)); };
  usb.onEndpointWrite = [this](uint32_t ep, const std::vector<uint8_t> &buf) {
    if (ep == 0 && buf.empty()) {
      if (!descriptorsSize.has_value()) {
        this->usb.sendSetupPacket(getDescriptorPacket(DescriptorType::Configration, CONFIGURATION_DESCRIPTOR_SIZE));
      } else if (!initialized) {
        // configured: port 0 opens, as USBCDC's one port does
        initialized = true;
        busy = false;
        setLines(0, CDC_DTR | CDC_RTS);
        if (onDeviceConnected) onDeviceConnected();
      } else {
        // a control request's status stage: the next may go
        busy = false;
        next();
      }
    }
    if (ep == 0 && buf.size() > 1) {
      if (buf.size() == CONFIGURATION_DESCRIPTOR_SIZE &&
          buf[1] == static_cast<uint32_t>(DescriptorType::Configration) && !descriptorsSize.has_value()) {
        descriptorsSize = (static_cast<uint32_t>(buf[3]) << 8) | buf[2];
        this->usb.sendSetupPacket(getDescriptorPacket(DescriptorType::Configration, *descriptorsSize));
      } else if (descriptorsSize.has_value() && descriptors.size() < *descriptorsSize) {
        descriptors.insert(descriptors.end(), buf.begin(), buf.end());
      }
      if (descriptorsSize.has_value() && *descriptorsSize == descriptors.size()) {
        const auto found = extractPorts(descriptors);
        for (size_t i = 0; i < found.size() && i < ports.size(); i++) {
          ports[i].comm = found[i].comm;
          ports[i].in = found[i].in;
          ports[i].out = found[i].out;
        }
        this->usb.sendSetupPacket(setDeviceConfigurationPacket(1));
      }
    }
    if (ep != 0) {
      for (auto &p : ports) {
        if (p.in == static_cast<int32_t>(ep)) {
          if (p.onSerialData) p.onSerialData(buf);
          break;
        }
      }
    }
  };
  usb.onEndpointRead = [this](uint32_t ep, uint32_t size) {
    for (auto &p : ports) {
      if (p.out != static_cast<int32_t>(ep)) continue;
      std::vector<uint8_t> b(std::min(size, p.txFIFO.itemCount()));
      for (auto &x : b) x = static_cast<uint8_t>(p.txFIFO.pull());
      this->usb.endpointReadDone(ep, std::move(b));
      return;
    }
  };
}

size_t CdcHost::portCount() const {
  return static_cast<size_t>(std::count_if(ports.begin(), ports.end(), [](const Port &p) { return p.in >= 0; }));
}

void CdcHost::setLines(size_t p, uint32_t value) {
  pending.emplace_back(p, value);
  next();
}

void CdcHost::next() {
  if (busy || !initialized || pending.empty()) return;
  busy = true;
  const auto [p, value] = pending.front();
  pending.pop_front();
  usb.sendSetupPacket(createSetupPacket({
      DataDirection::HostToDevice,
      SetupType::Class,
      SetupRecipient::Interface,
      CDC_REQUEST_SET_CONTROL_LINE_STATE,
      value,
      static_cast<uint32_t>(ports[p].comm) & 0xffff,
      0,
  }));
}

}  // namespace rp2040js
